### PR TITLE
Introduce users and authentication

### PR DESCRIPTION
**Pull request by cpytel**
* Introduces users and authentication using Clearance.
* Campaigns belong to users
* Players do not need to be logged in to view a campaign

### CODE REVIEW

File: .sample.env [hunk 1]
```diff
@@ -0,0 +1 @@
+HOST=yourapp.heroku.com
```
Comment (cpytel) [inline .sample.env:1]: We need to change this to something real.

File: lib/tasks/dev/prime.rake [hunk 2]
```diff
@@ -1,7 +1,19 @@
 namespace :dev do
   desc "Sample data for local development environment"
   task prime: :environment do
-    campaign = Campaign.find_or_create_by(name: "Dragon of Icespire Peak")
+    dm = User.find_or_initialize_by(
+      email: "jdbann@icloud.com",
```
Comment (cpytel) [inline lib/tasks/dev/prime.rake:5]: @jdbann let me know if this isn't the right email or if you'd prefer not to have it here at all and we can find another way to do this...

File: db/migrate/20200404113617_add_users_to_campaigns.rb [hunk 3]
```diff
@@ -0,0 +1,5 @@
+class AddUsersToCampaigns < ActiveRecord::Migration[6.0]
+  def change
+    add_belongs_to :campaigns, :user, null: true
```
Comment (cpytel) [inline db/migrate/20200404113617_add_users_to_campaigns.rb:3]: After this is run on the server and existing campaigns are associated with a user, we can add a new migration to change this to not allow null.

File: lib/tasks/dev/prime.rake [hunk 4]
```diff
@@ -1,7 +1,26 @@
...
+    )
+    unless dm.persisted?
+      dm.password = "prime"
+      dm.save
+    end
+
+    campaign = Campaign.find_by(
+      name: "Dragon of Icespire Peak"
+    )
+    if campaign
```
Comment (cpytel) [inline lib/tasks/dev/prime.rake:16]: This needs to be written this way right now to account for existing campaigns not owned by a dm.

Once this has been run, we can change to just

```
campaign = Campaign.find_or_create_by(
  name: "Dragon of Icespire Peak",
  user: dm
)
```

File: .sample.env (hunk 1 above)
Comment (jdbann) [reply .sample.env:1]: I don't follow. A real domain name?

File: lib/tasks/dev/prime.rake (hunk 2 above)
Comment (jdbann) [reply lib/tasks/dev/prime.rake:5]: I don't mind, but maybe something more generic like `dm@table.com` would be easier for other contributors to remember?

File: db/migrate/20200404113617_add_users_to_campaigns.rb (hunk 3 above)
Comment (jdbann) [reply db/migrate/20200404113617_add_users_to_campaigns.rb:3]: I can't see a way to invite you to the whole pipeline on heroku - do I have to create a new team for that instead?

File: app/views/layouts/application.html.erb [hunk 5]
```diff
@@ -12,15 +12,22 @@
   <body>
     <div class="bg-gray-200">
       <div class="container mx-auto py-4">
-        <div>
-          <%= link_to root_path do %>
+        <div class="flex items-stretch">
```
Comment (jdbann) [inline app/views/layouts/application.html.erb:15]: Could we go with `items-baseline` here?

File: config/initializers/clearance.rb [hunk 6]
```diff
@@ -0,0 +1,5 @@
+Clearance.configure do |config|
+  config.mailer_sender = "reply@example.com"
```
Comment (jdbann) [inline config/initializers/clearance.rb:2]: Can we load this from the environment?

File: db/migrate/20200329191558_create_users.rb [hunk 7]
```diff
@@ -0,0 +1,15 @@
+class CreateUsers < ActiveRecord::Migration[6.0]
+  def change
+    create_table :users do |t|
```
Comment (jdbann) [inline db/migrate/20200329191558_create_users.rb:3]: For all other records, we're using `id: :uuid`. Do you know if this is supported by Clearance?

**Comment by jdbann** [comment]
Thanks @cpytel! I've only got a few minor comments but I really appreciate this!

File: .sample.env (hunk 1 above)
Comment (cpytel) [reply .sample.env:1]: Yes, either the Heroku app’s domain or we’ll want to register a domain of our own.

File: app/views/layouts/application.html.erb (hunk 5 above)
Comment (cpytel) [reply app/views/layouts/application.html.erb]: Yup, done in 1ab4145

File: .sample.env (hunk 1 above)
Comment (jdbann) [reply .sample.env:1]: I've reserved https://table-production.herokuapp.com for now. Any thoughts on a domain name? Table is just a name for the project, no idea what we should really call this!

File: db/migrate/20200329191558_create_users.rb (hunk 7 above)
Comment (cpytel) [reply db/migrate/20200329191558_create_users.rb:3]: I had assumed, incorrectly it turns out, that the default had been changed to uuid and didn't need to be specified. Any objection to me changing the config so the migration generator default to `id: :uuid`?

Comment (jdbann) [reply db/migrate/20200329191558_create_users.rb:3]: Hmm, it has been set https://github.com/jdbann/table/blob/master/config/application.rb#L21

Perhaps Clearance doesn't follow this?

Comment (cpytel) [reply db/migrate/20200329191558_create_users.rb]: Fixed in b3d1ec5, yes Clearance copies the migration in and doesn't use the generator.

File: db/migrate/20200404113617_add_users_to_campaigns.rb (hunk 3 above)
Comment (cpytel) [reply db/migrate/20200404113617_add_users_to_campaigns.rb]: I'm honestly not sure. What if you add me to the app as a collaborator via the commandline? https://devcenter.heroku.com/articles/collaborating#collaborator-permissions-for-apps-in-a-personal-account

Comment (jdbann) [reply db/migrate/20200404113617_add_users_to_campaigns.rb]: I've created a team and invited you, the collaborators would only be per-app which wouldn't be helpful if you wanted to do anything with review apps

**Comment by cpytel** [comment]
Just a status update on this. There are several linting errors remaining for the new linters that were added. I've fixed some of them, but I have to go do some family stuff now. I will hopefully be able to finish it up and get this merged later on this afternoon.

**Comment by jdbann** [comment]
> There are several linting errors remaining for the new linters that were added.

Feel free to disable or amend any. They're my slightly tweaked variation on the thoughtbot RuboCop settings.

**Pull request merged**